### ISSUE TITLE
distsql: IFNULL() does not work

### ISSUE THREAD

**Issue by masayoshi-louis** [cockroachdb/cockroach#15686]
1. Please supply the header (i.e. the first few lines) of your most recent
   log file **for each node in your cluster**. On most unix-based systems
   running with defaults, this boils down to the output of

I170504 10:57:13.979320 1 util/log/clog.go:990  [config] file created at: 2017/05/04 10:57:13
I170504 10:57:13.979320 1 util/log/clog.go:990  [config] running on machine: db1
I170504 10:57:13.979320 1 util/log/clog.go:990  [config] binary: CockroachDB CCL v1.0-rc.1-dirty (linux amd64, built 2017/05/01 18:33:34, go1.8.1)
I170504 10:57:13.979320 1 util/log/clog.go:990  [config] arguments: [cockroach start --host=db1 --locality=country=china,region=north,datacenter=ch-north-bj-1 --http-port=28080 --certs-dir=/var/lib/cockroach/certs --store=/var/lib/cockroach/data,attrs=ssd --cache=4GB]
I170504 10:57:13.979320 1 util/log/clog.go:990  line format: [IWEF]yymmdd hh:mm:ss.uuuuuu goid file:line msg utf8=✓
I170504 10:57:13.979318 1 cli/start.go:564  CockroachDB CCL v1.0-rc.1-dirty (linux amd64, built 2017/05/01 18:33:34, go1.8.1)
I170504 10:57:14.080903 61 cli/start.go:337  starting cockroach node
I170504 10:57:14.089390 61 storage/engine/rocksdb.go:380  opening rocksdb instance at "/var/lib/cockroach/data"
I170504 10:57:14.662181 61 server/config.go:413  1 storage engine initialized

I170502 12:16:21.336666 1 util/log/clog.go:990  [config] file created at: 2017/05/02 12:16:21
I170502 12:16:21.336666 1 util/log/clog.go:990  [config] running on machine: db2
I170502 12:16:21.336666 1 util/log/clog.go:990  [config] binary: CockroachDB CCL v1.0-rc.1-dirty (linux amd64, built 2017/05/01 18:33:34, go1.8.1)
I170502 12:16:21.336666 1 util/log/clog.go:990  [config] arguments: [cockroach start --host=db2 --locality=country=china,region=north,datacenter=ch-north-bj-1 --http-port=28080 --certs-dir=/var/lib/cockroach/certs --store=/var/lib/cockroach/data,attrs=ssd --join=db1:26257 --cache=4GB]
I170502 12:16:21.336666 1 util/log/clog.go:990  line format: [IWEF]yymmdd hh:mm:ss.uuuuuu goid file:line msg utf8=✓
I170502 12:16:21.336665 1 cli/start.go:564  CockroachDB CCL v1.0-rc.1-dirty (linux amd64, built 2017/05/01 18:33:34, go1.8.1)
I170502 12:16:21.438097 21 cli/start.go:337  starting cockroach node
W170502 12:16:21.446876 21 gossip/gossip.go:1189  [n?] no incoming or outgoing connections
I170502 12:16:21.447107 21 storage/engine/rocksdb.go:380  opening rocksdb instance at "/var/lib/cockroach/data"
I170502 12:16:21.475686 57 gossip/client.go:131  [n?] started gossip client to db1:26257

I170502 12:16:27.985015 1 util/log/clog.go:990  [config] file created at: 2017/05/02 12:16:27
I170502 12:16:27.985015 1 util/log/clog.go:990  [config] running on machine: db3
I170502 12:16:27.985015 1 util/log/clog.go:990  [config] binary: CockroachDB CCL v1.0-rc.1-dirty (linux amd64, built 2017/05/01 18:33:34, go1.8.1)
I170502 12:16:27.985015 1 util/log/clog.go:990  [config] arguments: [cockroach start --host=db3 --locality=country=china,region=north,datacenter=ch-north-bj-1 --http-port=28080 --certs-dir=/var/lib/cockroach/certs --store=/var/lib/cockroach/data,attrs=ssd --join=db1:26257 --cache=4GB]
I170502 12:16:27.985015 1 util/log/clog.go:990  line format: [IWEF]yymmdd hh:mm:ss.uuuuuu goid file:line msg utf8=✓
I170502 12:16:27.985014 1 cli/start.go:564  CockroachDB CCL v1.0-rc.1-dirty (linux amd64, built 2017/05/01 18:33:34, go1.8.1)
I170502 12:16:28.086261 68 cli/start.go:337  starting cockroach node
W170502 12:16:28.093431 68 gossip/gossip.go:1189  [n?] no incoming or outgoing connections
I170502 12:16:28.094208 68 storage/engine/rocksdb.go:380  opening rocksdb instance at "/var/lib/cockroach/data"
I170502 12:16:28.123571 21 gossip/client.go:131  [n?] started gossip client to db1:26257

2. Please describe the issue you observed:

- What did you do?

CREATE TABLE testnull (k INT PRIMARY KEY, v INT);
SELECT IFNULL(SUM(v), 0) FROM testnull;

- What did you expect to see?

+-------------------+
| IFNULL(sum(v), 0) |
+-------------------+
|                 0 |
+-------------------+

- What did you see instead?

+-------------------+
| IFNULL(sum(v), 0) |
+-------------------+
| NULL              |
+-------------------+

I can workaround this by disabling distsql.

**Comment by RaduBerinde** [comment]
This is the [plan](https://cockroachdb.github.io/distsqlplan/decode.html?eJykkM9qhDAQxu99CvlOLsxh9ejJQlsQtrbY3VPxkJpBhGwikwgti-9eNBS2Hkphj_P9-4VcYJ3mWp3Zo3hHhpYwiuvYeyeLFAOV_kSxJwx2nMIit4TOCaO4IAzBMAoc1YfhhpVmAUFzUINZR0cZzkq-ysA-2MkYEBq2mqVIyjzZo50Jbgo_yzP9n3nf98K9Cm6DfDs9p2W2A6F6eKyPaZnvrqjVU306HNIyo2QxbuA37EdnPf-ib_daAuue4296N0nHr-K6dTyeL2t6FTT7EN0sHpWN1vKs63L2ZznflNv57jsAAP__D0-g2A==).

This is a consequence of the way we generate pre- and post- renders (we pre-render them and then use IDENT "aggregators" instead of just using the constants directly in the post-render).
In this case, because there are no rows in the table, the constant is not passed through IDENT.

Issue #12525 tracks reworking the renders.

The non-dist sql code has a hack for this:
https://github.com/cockroachdb/cockroach/blob/8568e9a6732ed25cbb7821614b35b1dad4abbd9b/pkg/sql/group.go#L774

We can't really implement the same hack in distsql because the aggregation processor doesn't have access to the value.

CC @knz

**Comment by RaduBerinde** [comment]
@masayoshi-louis thanks for the detailed report! A workaround that doesn't involve disabling distsql is doing this:
```sql
SELECT IFNULL(s, 0) FROM (SELECT sum(v) AS s FROM testnull);
```
This does that the original query *should* be doing ([plan](https://cockroachdb.github.io/distsqlplan/decode.html?eJykkDFr8zAQhvfvV5h3SkBDnFGTw9cWDE5S3GYqHlTrMAZHMqcTtAT_92KLQuuhFDree_c-j9ANzls6mSsF6BfkaBRG9i2F4HmO0kFp36B3Cr0bo8xxo9B6JugbpJeBoPFsXgeqyVhiKFgS0w8LdOT-avi9EAri4jBA4RxFZ8UezaTgo3xCJ_V73aHrmDojfmV7uhw3Rb6FQk3OEuusfDhdqmpT5Crbaa3v7v-Xx0O1_Yu8pjB6F-ibes1rFMh2lH4x-MgtPbJvF3gaz8v1ElgKkrZ5GkqXVvOzvpbzH8v7VbmZ_n0EAAD___qJnxQ=)).

**Comment by RaduBerinde** [comment]
It looks like we can remove the unnecessary ident aggregations without doing all the refactoring for #12525. Hopefully we can get this fix in for 1.0.
